File: python/evaluation.py
#!/usr/bin/env python3

import math
import shutil
from typing import List
from typing import Optional  # pylint: disable=unused-import
from typing import cast

from bindings import Execution
from python.dispatcher import Dispatcher
from python.dispatcher import Event
from python.dispatcher import EventStatus
from python.source_file import SourceFile
from python.task import Subtask
from python.task import Task
from python.task import Testcase
from python.ui import EvaluationResult
from python.ui import EvaluationStatus
from python.ui import UI


class SingleEvaluationState:
    def __init__(self, execution: Execution, check: Execution,
                 subtask_num: int, testcase_offset: int,
                 has_checker: bool) -> None:
        self.execution = execution
        self.check = check
        self.subtask_num = subtask_num
        self.testcase_offset = testcase_offset
        self.has_checker = has_checker


class SubtaskScoreInfo:
    def __init__(self, evaluation: 'Evaluation', subtask: Subtask,
                 ui: UI) -> None:
        self._subtask = subtask
        self._testcase_scores = \
            [None for _ in subtask.testcases]  # type: List[Optional[float]]
        self._evaluation = evaluation
        self._ui = ui

    def update_score(self, testcase_offset: int, score: float) -> None:
        self._testcase_scores[testcase_offset] = score
        if all(score is not None for score in self._testcase_scores):
            testcase_scores = cast(List[float], self._testcase_scores)
            score = self._subtask.compute_score(testcase_scores)
            self._ui.set_subtask_score(self._subtask.num,
                                       self._evaluation.solution_src, score)
            self._evaluation.update_score(self._subtask.num, score)


class Evaluation:
    def _callback(self, testcase_num: int, event: Event,
                  status: EventStatus) -> bool:
        evaluation_state = self._evaluations[testcase_num]
        execution = evaluation_state.execution

        def evaluation_result(score: float, msg: str) -> EvaluationResult:
            return EvaluationResult(score, msg,
                                    execution.cpu_time(),
                                    execution.wall_time(), execution.memory())

        if status == EventStatus.START:
            if execution.id() == event.id():
                self._ui.set_evaluation_status(testcase_num, self.solution_src,
                                               EvaluationStatus.EXECUTING)
            else:
                self._ui.set_evaluation_status(testcase_num, self.solution_src,
                                               EvaluationStatus.CHECKING)
            return True
        if status == EventStatus.FAILURE and evaluation_state.has_checker \
                and event.id() == evaluation_state.check.id():
            display_msg = evaluation_state.check.stderr().contents(1024 * 1024)
            self._ui.set_evaluation_status(testcase_num, self.solution_src,
                                           EvaluationStatus.FAILURE,
                                           evaluation_result(0.0, display_msg))
            return False
        if event.id() == execution.id():
            if execution.cpu_time() <= self._task.time_limit:
                if status == EventStatus.SUCCESS:
                    self._ui.set_evaluation_status(testcase_num,
                                                   self.solution_src,
                                                   EvaluationStatus.EXECUTED)
                    return True
                if execution.signal() != 0:
                    display_msg = "Signal " + str(execution.signal())
                elif execution.status_code() != 0:
                    display_msg = "Return code " + str(execution.status_code())
                else:
                    display_msg = "Missing output files"
            else:
                display_msg = "Execution timed out"
            score = 0.0
        else:
            if not evaluation_state.has_checker:
                if status == EventStatus.FAILURE:
                    display_msg = "Output not correct"
                    score = 0.0
                else:
                    display_msg = "Output is correct"
                    score = 1.0
            else:
                display_msg = evaluation_state.check.stderr().contents(
                    1024 * 1024)
                try:
                    score = float(
                        evaluation_state.check.stdout().contents(1024 * 1024))
                except ValueError:
                    self._ui.set_evaluation_status(
                        testcase_num, self.solution_src,
                        EvaluationStatus.FAILURE,
                        evaluation_result(0.0, display_msg),
                        "Invalid score returned by checker")
        self._ui.set_evaluation_status(testcase_num, self.solution_src,
                                       EvaluationStatus.SUCCESS,
                                       evaluation_result(score, display_msg))
        self._subtask_score_info[evaluation_state.subtask_num].update_score(
            evaluation_state.testcase_offset, score)
        return True

    def _evaluate_testcase(self, num: int, testcase: Testcase, exclusive: bool,
                           cache_mode: Execution.CachingMode,
                           eval_executor: Optional[str],
                           extra_eval_time: float) -> None:
        def callback(event: Event, status: EventStatus) -> bool:
            return self._callback(num, event, status)

        if testcase.input_id is None or testcase.output_id is None \
                or testcase.subtask is None:
            raise ValueError("Invalid testcase state")

        execution = self._solution.execute(
            "Evaluation of solution %s on testcase %d" %
            (self.solution_src, num), [], callback, exclusive, cache_mode)
        if eval_executor is not None:
            execution.set_executor(eval_executor)
        # CPU time can only be set to an integer
        execution.cpu_limit(self._task.time_limit + math.ceil(extra_eval_time))
        execution.wall_limit(self._task.time_limit + extra_eval_time)
        execution.memory_limit(self._task.memory_limit)
        contestant_output = self._task.setup_io(execution, testcase.input_id)
        check_description = "Checking result of solution %s on testcase %d" % (
            self.solution_src, num)
        if self._task.checker is None:
            has_checker = False
            # TODO(veluca): replace this with our own utility?
            check = self._dispatcher.add_execution(
                check_description,
                cast(str, self._diff_path),
                ["-w", "output", "contestant_output"],
                callback,
                exclusive=False,
                cache_mode=cache_mode)
        else:
            has_checker = True
            check = self._task.checker.execute(
                check_description, ["input", "output", "contestant_output"],
                callback,
                exclusive=False,
                cache_mode=cache_mode)
            check.input("input", testcase.input_id)
        check.input("output", testcase.output_id)
        check.input("contestant_output", contestant_output)
        check.cpu_limit(10 * self._task.time_limit)
        check.memory_limit(10 * self._task.memory_limit)
        testcase_offset = testcase.num - testcase.subtask.testcases[0].num
        self._evaluations.append(
            SingleEvaluationState(execution, check, testcase.subtask.num,
                                  testcase_offset, has_checker))
        self._ui.set_evaluation_status(num, self.solution_src,
                                       EvaluationStatus.WAITING)

    def __init__(self, dispatcher: Dispatcher, ui: UI, task: Task,
                 solution: str, exclusive: bool,
                 eval_cache_mode: Execution.CachingMode,
                 eval_executor: Optional[str], extra_eval_time: float) -> None:
        if not task.generated:
            raise ValueError("You must first generate the task")
        self._diff_path = shutil.which("diff")
        if task.checker is None and self._diff_path is None:
            raise RuntimeError(
                "Could not find diff utility and no checker present")
        self.solution_src = solution
        self.subtask_scores = \
            [None for _ in task.subtasks]  # type: List[Optional[float]]
        self.score = None  # type: Optional[float]
        self._dispatcher = dispatcher
        self._solution = SourceFile(dispatcher, ui, solution, is_solution=True)
        self._solution.compile(
            task.graders(self._solution.get_language()), eval_cache_mode)
        self._task = task
        self._ui = ui
        self._evaluations = []  # type: List[SingleEvaluationState]
        self._subtask_score_info = \
            [SubtaskScoreInfo(self, subtask, ui) for subtask in task.subtasks]
        for num, testcase in enumerate(task.testcases):
            self._evaluate_testcase(num, testcase, exclusive, eval_cache_mode,
                                    eval_executor, extra_eval_time)

    def update_score(self, subtask_num: int, score: float) -> None:
        self.subtask_scores[subtask_num] = score
        if all(score is not None for score in self.subtask_scores):
            self.score = sum(self.subtask_scores)
            self._ui.set_task_score(self.solution_src, cast(float, self.score))

File: manager/evaluation.cpp
#include "manager/evaluation.hpp"
#include "glog/logging.h"

namespace manager {

Evaluation::Evaluation(EventQueue* queue, core::Core* core,
                       const Generation& generation, const proto::Task& task,
                       bool exclusive, proto::CacheMode cache_mode,
                       const std::string& executor)
    : generation_(generation) {
  queue_ = queue;
  core_ = core;
  task_ = task;
  exclusive_ = exclusive;
  cache_mode_ = cache_mode;

  for (auto subtask : task.subtasks()) {
    for (auto testcase : subtask.second.testcases()) {
      testcases_of_subtask[subtask.first].push_back(testcase.first);
    }
  }
}

void Evaluation::Evaluate(SourceFile* solution) {
  LOG(INFO) << "Evaluating " << solution->Name();
  std::string name = solution->Name();
  status_[name].task_score = -1;  // assuming you cannot make negative points

  for (auto subtask : task_.subtasks()) {
    for (auto testcase : subtask.second.testcases()) {
      evaluate_testcase_(subtask.first, testcase.first, solution);
    }
  }
}

void Evaluation::evaluate_testcase_(int64_t subtask_num, int64_t testcase_num,
                                    SourceFile* solution) {
  std::string name = solution->Name();
  std::string s_testcase_num = std::to_string(testcase_num);

  core::Execution* execution = solution->execute(
      "Evaluation of " + name + " on testcase " + s_testcase_num, {});
  execution->CpuLimit(task_.time_limit());
  execution->WallLimit(task_.time_limit() * 1.5f);
  execution->MemoryLimit(task_.memory_limit_kb());
  if (exclusive_) execution->SetExclusive();
  if (cache_mode_ == proto::ALL)
    execution->SetCachingMode(core::Execution::SAME_EXECUTOR);
  else
    execution->SetCachingMode(core::Execution::NEVER);
  if (!executor_.empty()) execution->SetExecutor(executor_);
  // setup solution i/o
  if (task_.input_file().empty())
    execution->Stdin(generation_.GetInput(testcase_num));
  else
    execution->Input(task_.input_file(), generation_.GetInput(testcase_num));
  core::FileID* output;
  if (task_.output_file().empty())
    output = execution->Stdout();
  else
    output = execution->Output(task_.output_file());
  execution->SetCallback([this, name, testcase_num,
                          subtask_num](const core::TaskStatus& status) -> bool {
    if (status.type == core::TaskStatus::FILE_LOAD)
      return !(status.event == core::TaskStatus::FAILURE);
    if (status.event == core::TaskStatus::START)
      queue_->Executing(name, testcase_num);
    if (status.event == core::TaskStatus::SUCCESS) {
      auto exec = status.execution_info;
      if (exec->Success()) {
        queue_->Executed(name, testcase_num);
      } else {
        queue_->EvaluationDone(name, testcase_num, 0.0, exec->Message(),
                               exec->CpuTime(), exec->WallTime(),
                               exec->Memory());
        UpdateScore(name, subtask_num, testcase_num, 0.0);
      }
    }
    if (status.event == core::TaskStatus::FAILURE) {
      queue_->FatalError(status.message + ": " +
                         status.execution_info->Message());
      return false;
    }
    return true;
  });

  core::Execution* checker;
  if (task_.has_checker()) {
    checker = generation_.GetChecker()->execute(
        "Checking solution " + name + " for testcase " + s_testcase_num,
        {"input", "output", "contestant_output"});
    checker->Input("input", generation_.GetInput(testcase_num));
  } else {
    // TODO(veluca): replace this with our own utility?
    checker = core_->AddExecution(
        "Checking solution " + name + " for testcase " + s_testcase_num,
        // TODO(edomora97) getting the real path of diff is hard (which command)
        "/usr/bin/diff", {"-w", "output", "contestant_output"});
  }
  checker->Input("output", generation_.GetOutput(testcase_num));
  checker->Input("contestant_output", output);
  checker->CpuLimit(10 * task_.time_limit());
  checker->MemoryLimit(10 * task_.memory_limit_kb());
  if (cache_mode_ == proto::ALL)
    checker->SetCachingMode(core::Execution::ALWAYS);
  else
    checker->SetCachingMode(core::Execution::NEVER);
  if (!executor_.empty()) checker->SetExecutor(executor_);
  checker->SetCallback([this, name, subtask_num, testcase_num, checker,
                        execution](const core::TaskStatus& status) -> bool {
    if (status.type == core::TaskStatus::FILE_LOAD)
      return !(status.event == core::TaskStatus::FAILURE);
    if (status.event == core::TaskStatus::START) {
      queue_->Checking(name, testcase_num);
      return true;
    }

    if (!task_.has_checker()) {
      // without the checker diff is used to check and it exits with (1) if the
      // files differ.
      if (status.execution_info->Success()) {
        queue_->EvaluationDone(name, testcase_num, 1.0f, "Output is correct",
                               execution->CpuTime(), execution->WallTime(),
                               execution->Memory());
        UpdateScore(name, subtask_num, testcase_num, 1.0);
      } else {
        queue_->EvaluationDone(name, testcase_num, 0.0f,
                               "Output is not correct", execution->CpuTime(),
                               execution->WallTime(), execution->Memory());
        UpdateScore(name, subtask_num, testcase_num, 0.0);
      }
    } else {
      if (status.execution_info->Success()) {
        std::string out = checker->Stdout()->Contents(1024 * 1024);
        std::string err = checker->Stderr()->Contents(1024 * 1024);
        try {
          float score = std::stof(out);
          if (score < 0.0 || score > 1.0)
            throw std::invalid_argument("score not in range [0.0, 1.0]");
          queue_->EvaluationDone(name, testcase_num, score, err,
                                 execution->CpuTime(), execution->WallTime(),
                                 execution->Memory());
          UpdateScore(name, subtask_num, testcase_num, score);
        } catch (const std::invalid_argument& ex) {
          queue_->EvaluationFailure(name, testcase_num, "checker failed",
                                    execution->CpuTime(), execution->WallTime(),
                                    execution->Memory());
          queue_->FatalError(std::string("Checker returned invalid score: ") +
                             ex.what() + "\n" + "The stdout was: " + out +
                             "\n" + "The stderr was: " + err);
          return false;
        }
      } else {
        std::string checker_error = checker->Stderr()->Contents(1024 * 1024);
        queue_->EvaluationFailure(
            name, testcase_num, checker->Message() + "\n" + checker_error,
            execution->CpuTime(), execution->WallTime(), execution->Memory());
        queue_->FatalError("Checker failed: " + checker_error);
        return false;
      }
    }
    return true;
  });

  queue_->EvaluationWaiting(name, testcase_num);
}
void Evaluation::UpdateScore(const std::string& name, int64_t subtask_num,
                             int64_t testcase_num, float score) {
  Evaluation::EvaluationStatus& status = status_[name];
  proto::ScoreMode score_mode = task_.subtasks().at(subtask_num).score_mode();
  status.testcase_scores[testcase_num] = score;
  for (int64_t testcase : testcases_of_subtask[subtask_num])
    if (status.testcase_scores.count(testcase) == 0) return;
  float st_score = 0;
  std::function<float(float, float)> acc = nullptr;
  switch (score_mode) {
    case proto::MIN:
      st_score = 1.0;
      acc = [](float score1, float score2) { return std::min(score1, score2); };
      break;
    case proto::MAX:
      st_score = 0.0;
      acc = [](float score1, float score2) { return std::max(score1, score2); };
      break;
    case proto::SUM:
      st_score = 0.0;
      acc = [](float score1, float score2) { return score1 + score2; };
      break;
    default:
      throw std::domain_error("Invalid score mode for subtask " +
                              std::to_string(subtask_num));
  }
  for (int64_t testcase : testcases_of_subtask[subtask_num])
    if (status.testcase_scores.count(testcase) > 0)
      st_score = acc(st_score, status.testcase_scores[testcase]);
  st_score *= task_.subtasks().at(subtask_num).max_score();
  if (score_mode == proto::SUM)
    st_score /= testcases_of_subtask[subtask_num].size();

  if (status.subtask_scores.count(subtask_num) == 0 ||
      st_score != status.subtask_scores.at(subtask_num))
    queue_->SubtaskTaskScore(name, st_score, subtask_num);
  status.subtask_scores[subtask_num] = st_score;

  float task_score = 0;
  for (auto subtask : status.subtask_scores) task_score += subtask.second;
  if (task_score != status.task_score) queue_->TaskScore(name, task_score);
  status.task_score = task_score;
}
}  // namespace manager
